### PR TITLE
Add jump class, update interpreter

### PR DESCRIPTION
**Pull request by jacques-cadet**
Just thought I'd add the JUMP class myself to complete the jump process.

### CODE REVIEW

File: exa.py [hunk 1]
```diff
@@ -54,6 +54,19 @@ def do(self, interp_state):
...
+class JUMP(Statement):
+
+    _expected_args = 2
+
+    def __init__(self, line_num, tokens):
+        super().__init__(line_num, tokens)
+        self._label = tokens[1]
+
+    def do(self, interp_state):
+        label_address = interp_state.labels[self._label]
```
Comment (dhellmann) [inline exa.py]: In most of the rest of the places where errors can occur I tried to produce an error message that showed the line number of the input that has the bad value. We should catch the KeyError here and produce a ValueError or RuntimeError.

File: exa.py [hunk 2]
```diff
@@ -125,6 +138,36 @@ def do(self, interp_state):
         interp_state.next_statement += 1
 
 
+class ADDI(MathStatement):
```
Comment (dhellmann) [inline exa.py:144]: I don't mind the cleanup changes being part of this PR, but please put them into separate commits.

File: exa.py [hunk 3]
```diff
@@ -259,6 +273,7 @@ def run(self, statements):
 
     def parse(self, statements):
         # eliminate blank lines
+        print(statements)
```
Comment (dhellmann) [inline exa.py]: I don't think we want this debug to go in, do we?

File: exa.py [hunk 4]
```diff
@@ -294,6 +312,5 @@ def parse(self, statements):
     with open(args.file, 'r') as f:
         statements = f.readlines()
 
-    interp = Interpreter()
-    result = interp.run(statements)
+    result = Interpreter().run(statements)
```
Comment (dhellmann) [inline exa.py]: I'm not sure why you changed these lines?

**Review by dhellmann — CHANGES_REQUESTED** [review]

File: exa.py (hunk 4 above)
Comment (jacques-cadet) [reply exa.py]: I'll make the changes above,  as for the clean up I did that earlier and forgot about it, will do. As for interp, this instance object isn't used anywhere else to access any other methods of Interpreter class so I just removed it. My apologies.

Comment (dhellmann) [reply exa.py]: No need to apologize, I was just curious. It's a style difference and I'm not sure I can explain why I like having the interpreter reference in its own variable.

Note I've pushed some changes this afternoon that mean you'll need to rebase this and update the init() method of the new class.

**Comment by jacques-cadet** [comment]
Will do, here's a snippet of the EXA challenge regarding the file handling.

Challenge 4: File handling
<https://github.com/jacques-cadet/jam-sessions/tree/master/2019-08#instructions-2>
Instructions

All that programming is no good if it cannot input and output. Thankfully,
EXAs can manipulate files and leave traces.

In our simplified emulator, a file is simply a list of numbers with a name.
To use a file, an EXA must first grab that file by its name. Once done, it
has to drop it before using a different file.

Here is also where the F register comes in play. The reference guide reads:

The F register allows an EXA to read and write the contents of a held file.
When an EXA grabs a file, its "file cursor" will be set to the first value
in the file. Reading from the F register will read this value; writing to
the F register will overwrite this value. After reading or writing the
F register,
the file cursor will automatically advance. Writing to the end of the file
will append a new value instead of overwriting.

*Note:* This is not the same file handling as the one you'd do with open in
Python. These files are virtual; lists which exist only in the emulator.
You will need to add them to your simulation.

The file manipulation instructions to implement are:

   - GRAB R/N
   Grab a file with the specified ID.
   - FILE R
   Copy the ID of the file into the specified register.
   - SEEK R/N
   Move the file cursor forward (positive) or backward (negative) by the
   specified number of values.
   If SEEK would move the file cursor past the beginning or end of the file
   it will instead be clamped. Thus, you can use values of -9999 or 9999 to
   reliably move to the beginning or end of a file.
   - VOID F
   Remove the value highlighted by the file cursor from the currently held
   file.
   - DROP
   Drop the currently held file.
   - TEST EOF
   If the file pointer is currently at the end of the held file, set
the T register
   to 1, otherwise set the T register to 0.

Note that trying to access this register when the EXA isn't holding a file
will result in an "Invalid F register access" error, causing the EXA to
crash and terminate. If an EXA tries to open a file that does not exist (or
is otherwise not available), it also crashes.

Modify your emulator to support file handling. On top of the code to run,
it will need to be initialized with files (and their contents). It will
need to return the file contents at the end of execution.
<https://github.com/jacques-cadet/jam-sessions/tree/master/2019-08#example-2>
Example

Let's look at a simple file handling program. In this scenario, the
environment starts with two existing files: one called 100 with a list of
numbers, and a second one called 200 that's empty.

GRAB 100
MARK FILE_READ
ADDI F X X
TEST EOF
FJMP FILE_READ
DROP
GRAB 200
COPY X F
DROP

This time, try to walk through the instructions yourself. This script reads
the numbers in 100, sums them, and writes the sum to 200.
<https://github.com/jacques-cadet/jam-sessions/tree/master/2019-08#final-challenge>Final
Challenge

Here's the final code I want you to test your emulator against. The virtual
world is expected to contain a file with the ID 400 that is initially empty.

GRAB 400
COPY 1 X
MARK A
SEEK -9999
ADDI X 1 X
TEST X < 50
FJMP D
MARK B
TEST EOF
TJMP C
MODI X F T
FJMP A
JUMP B
MARK C
COPY X F
JUMP A
MARK D
DROP

What are the contents of the file 400 at the end of the program? Can you
identify what calculation this program is doing?
